core/controllers/reader_test.py

```diff
@@ -204,6 +211,125 @@ def test_huge_answers(self):
             'a' * 1000500, 'Sorry, nope, we didn\'t get it', '')
 
 
+class ReaderClassifyTests(test_utils.GenericTestBase):
+    """Test reader.classify using the sample explorations.
+
+    Since the end to end tests cover correct classification,
+    ReaderClassifyTests is only checking which of the hard/soft/classifier
+    rules is classified on input.
+    """
+
+    def setUp(self):
+        super(ReaderClassifyTests, self).setUp()
+        self._init_classify_inputs('16')
+
+    def _init_classify_inputs(self, exploration_id):
+        test_exp_filepath = os.path.join(
+            feconf.TESTS_DATA_DIR, 'string_classifier_test.yaml')
+        yaml_content = utils.get_file_contents(test_exp_filepath)
+        assets_list = []
+        exp_services.save_new_exploration_from_yaml_and_assets(
+            feconf.SYSTEM_COMMITTER_ID, yaml_content,
+            'Testing String Classifier', 'Test',
+            exploration_id, assets_list)
+
+        self.EXP_ID = exploration_id
+        self.EXP_STATE = (
+            exp_services.get_exploration_by_id(exploration_id).states['Home'])
+
+    def _get_classiying_rule_type(self, answer):
+        string_classifier_ctor = classifier_services.StringClassifier.__init__
+        string_classifier_ctor_counter = test_utils.CallCounter(
+            string_classifier_ctor)
+
+        with self.swap(
+            classifier_services.StringClassifier,
+            '__init__', string_classifier_ctor_counter):
+            response = reader.classify(
+                self.EXP_ID, self.EXP_STATE, answer, {'answer': answer})
+
+        answer_group_index = response['answer_group_index']
+        rule_spec_index = response['rule_spec_index']
+        answer_groups = self.EXP_STATE.interaction.answer_groups
+        if answer_group_index == len(answer_groups):
+            return 'default'
+
+        answer_group = answer_groups[answer_group_index]
+        if answer_group.get_fuzzy_rule_index() == rule_spec_index:
+            return (
+                'classifier'
+                if string_classifier_ctor_counter.times_called > 0 else 'soft')
```

Comment:
If you switch to checking ``predict_label_for_doc`` it will probably be better to do a strict equivalence check, e.g. ``string_classifier_predict_counter.times_called == 0`` since it really should only use the classifier once.